core/src/sumeragi/network_topology.rs

```diff
@@ -390,8 +364,8 @@ impl Topology {
     }
 
     /// Config param telling topology when to reshuffle at view change.
-    pub const fn reshuffle_after(&self) -> u64 {
-        self.reshuffle_after_n_view_changes
+    pub fn reshuffle_after(&self) -> u64 {
+        self.sorted_peers.len() as u64
```

Comment:
I also think it's better not to have it be a configuration parameter.

Here are some of my preliminary thoughts.

I think you are assuming that the peers are shifted by one. I think they are but I'm not entirely certain. Can you confirm this?

Now, if we analyze the problem here, the end goal is to find the next valid topology, if there is a valid topology, in a minimum number of iterations in the average case (or maybe in the worst case?). Since we don't have hijiri yet and we assume all peers to be equally trustworthy I think we could gain some benefit from making it a bit more stochastic, i.e. reshuffling even before we've exhausted all the permutations with the current relative ordering.

Reply:
I also notice, that we don't bring down `view_change_proofs` to 0 after reshuffling. Is that ok? I think this indicates that we'll be reshuffling all the time after `reshuffle_after_n_view_changes` is reached

Reply:
> Can you confirm this?

Yeah, confirmed by a new unit test -- `network_topology::tests::topology_shifts_by_one_at_view_change`

> find the next valid topology

Interesting problem.
I guess one of your saying is that just shifting by one is very likely to result in similar faulty peers.
So if every peer were uniformly faulty, I'd minimize the overlapped section -- shift by f.
In reality, however, things are more complicated because
- the leader and the proxy tail are privileged
- there can be also artificial faults

> reshuffling all the time after `reshuffle_after_n_view_changes` is reached

You are right. This is a mis-implementation unless the parameter increases with every reshuffle.
https://github.com/hyperledger/iroha/blob/81f46bb1a14e16d6eb4cf33f691b673237ec371d/core/src/sumeragi/network_topology.rs#L208
Well, I think we can check if `view_change_proofs.len()` is congruent to 0 modulo `sorted_peers.len()`
-- as long as we take the current shift-by-one strategy

Reply:
>  Well, I think we can check if view_change_proofs.len() is congruent to 0 modulo sorted_peers.len()

yeah, this would do in the current implementation. If you're going to fix this I would appreciate a test as well, if not, please open an issue

Reply:
It's hard to say what would be the best method of shifting/shuffling peers. Maybe we should open a discussion about this in a separate issue and have it shift a full circle here as you did

Reply:
As for the shuffle judge I'll fix it and make a test.
As for the comprehensive discussion about the view change method, created https://github.com/hyperledger/iroha/issues/2133